checker/check.go

```diff
@@ -139,6 +147,15 @@ func checkProfile(p *proto.Profile, req *requestCache) (action Action) {
 }
 
 func checkRules(rules []*proto.Rule, req *requestCache, policyNamespace string) (action Action) {
+	defer func() {
+		// match checks data from dataplane which may be invalid.
+		// This is represented by the match function throwing a panic.
+		// Recover from the panic and perculate it up to the caller as INVALID_DATA
+		if r := recover(); r != nil {
+			log.Error("Match paniced; return INVALID_DATA")
+			action = INVALID_DATA
+		}
+	}()
```

Comment:
why are you recovering from the panic here, rather than in `Check()`?

It seems strange to recover from the panic, only to convert it to a different error condition which has to be handled again, higher up in the stack.  If you put the recovery into `Check()`, which is basically the highest level (that is still our code and not gRPC), then we can put it all in one place.

Reply:
doing a panic recovery in `Check()` is fine.
One thing to consider is that `Check()` calls `checkStore()` with a lock and we'd have to make sure we do unlocking in defer (since defer functions are invoked even with a `panic()` stack unwind).
If we recovered in `checkStore()` then we'd be doing recovery all within the context of the lock.
Something to consider. wdyt?
Meanwhile i'll code up recovery in `Check()`.
Rest of the comments are fine too and I'll address them with the next patch. Thanks.

Reply:
That's a good point. We don't unlock in `defer` but maybe we should anyway?  I don't really mind either way; `Check()` is a pretty light wrapper around `checkStore()`.

Reply:
i am going to change the unlocks to defer as it's safe to do that.